# Обязательные задачи

In [1]:
import matplotlib.pyplot as plt
import pandas as pd

from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import accuracy_score, confusion_matrix, mean_absolute_error
from sklearn.model_selection import train_test_split 
from sklearn.neural_network import MLPClassifier

In [2]:
df = pd.read_csv('vehicles_dataset_prepared.csv')
df.head()

id  price price_category  is_audi  is_ford  is_chevrolet  \
0  7308295377  54990           high        0        0             0   
1  7316380095  16942         medium        0        1             0   
2  7313733749  35590           high        0        0             0   
3  7308210929  14500         medium        0        0             0   
4  7303797340  14590         medium        0        0             0   

   is_toyota  x0_diesel  x0_electric  x0_gas  ...  long_std  year_std  \
0          0        1.0          0.0     0.0  ...  0.484245  1.322394   
1          0        0.0          0.0     0.0  ...  1.110800  0.695973   
2          0        0.0          0.0     1.0  ...  0.531185  0.852578   
3          1        0.0          0.0     1.0  ...  0.853562  0.226157   
4          0        0.0          0.0     0.0  ...  0.557607  0.069552   

   odometer/price_std  desc_len_std  model_in_desc_std  model_len_std  \
0           -0.510784      0.632075          -0.155788       1.163032   
1           -0.402947     -0.646781          -0.155788       0.932087   
2           -0.514480      0.560744          -0.155788       0.470197   
3           -0.241883      0.180435          -0.155788      -0.915473   
4           -0.333074      0.766366          -0.155788       1.163032   

   model_word_count_std  month_std  dayofweek_std  diff_years_std  
0              1.910669  -0.615846       1.120284       -1.322394  
1              1.235799   1.623784      -1.374972       -0.695973  
2              0.560930  -0.615846      -0.376870       -0.852578  
3             -0.788810  -0.615846       1.120284       -0.226157  
4              1.910669  -0.615846       0.122182       -0.069552  

[5 rows x 1463 columns]

# Задача 0. Линейная регрессия
Вспомните задачу по предсказанию стоимости поддержанного автомобиля. Попробуем обучить модель линейной регрессии для предсказания цены автомобиля (колонка price). Для этого проделайте следующие шаги:

Подготовьте данные: удалите колонки, которые косвенно содержат информацию о целевой переменной (odometer/price_std, price_category).
Разделите выборку на треин и тест в отношении 70/30.
Обучите модель линейной регрессии с дефолтными параметрами.
Посчитайте значение метрики mae на тестовой выборке для линейной регрессии.
Выведите получившиеся коэффициенты линейной регрессии при каждом параметре обучающей выборки с помощью метода coef_. Есть ли коэффициенты, которые равны нулю? Если есть, выведите названия фичей с нулевым коэффициентом
Удалите фичи, коэффициенты которых равны нулю. И переобучите модель. Убедитесь, что значение метрики не изменилось.

In [3]:
df_2 = df.copy()

In [4]:
df_2 = df_2.drop(columns = ['odometer/price_std','price_category'])
df_2.head()

id  price  is_audi  is_ford  is_chevrolet  is_toyota  x0_diesel  \
0  7308295377  54990        0        0             0          0        1.0   
1  7316380095  16942        0        1             0          0        0.0   
2  7313733749  35590        0        0             0          0        0.0   
3  7308210929  14500        0        0             0          1        0.0   
4  7303797340  14590        0        0             0          0        0.0   

   x0_electric  x0_gas  x0_hybrid  ...   lat_std  long_std  year_std  \
0          0.0     0.0        0.0  ... -0.619172  0.484245  1.322394   
1          0.0     0.0        0.0  ...  0.388014  1.110800  0.695973   
2          0.0     1.0        0.0  ... -0.843059  0.531185  0.852578   
3          0.0     1.0        0.0  ... -0.504509  0.853562  0.226157   
4          0.0     0.0        0.0  ... -0.460101  0.557607  0.069552   

   desc_len_std  model_in_desc_std  model_len_std  model_word_count_std  \
0      0.632075          -0.155788       1.163032              1.910669   
1     -0.646781          -0.155788       0.932087              1.235799   
2      0.560744          -0.155788       0.470197              0.560930   
3      0.180435          -0.155788      -0.915473             -0.788810   
4      0.766366          -0.155788       1.163032              1.910669   

   month_std  dayofweek_std  diff_years_std  
0  -0.615846       1.120284       -1.322394  
1   1.623784      -1.374972       -0.695973  
2  -0.615846      -0.376870       -0.852578  
3  -0.615846       1.120284       -0.226157  
4  -0.615846       0.122182       -0.069552  

[5 rows x 1461 columns]

In [5]:
train, test = train_test_split(df_2, test_size=0.3, random_state=42)

In [6]:
lr = LinearRegression()


In [7]:
lr.fit(train, train.price)

LinearRegression()

In [8]:
pred = lr.predict(test)
mean_absolute_error(test.price, pred)

4.686273083608963e-12

In [9]:
lr.coef_

array([-4.07478194e-20,  1.00000000e+00,  1.26371602e-12, ...,
       -3.56324865e-15, -5.00490228e-14,  4.62722605e-14])

In [10]:
coeff_df = pd.DataFrame(lr.coef_, df_2.columns, columns=['Коэффициент']) 
coeff_df


Коэффициент
id                   -4.074782e-20
price                 1.000000e+00
is_audi               1.263716e-12
is_ford               4.496038e-14
is_chevrolet          1.297979e-13
...                            ...
model_len_std        -9.232012e-14
model_word_count_std  1.172188e-13
month_std            -3.563249e-15
dayofweek_std        -5.004902e-14
diff_years_std        4.627226e-14

[1461 rows x 1 columns]

In [11]:
coeff_df.isnull().any()

Коэффициент    False
dtype: bool

# Задача 1. Логистическая регрессия
Теперь в рамках тех же данных попробуйте предсказать price_category с помощью алгоритма логистической регрессии. Предварительно из датафрейма удалите переменные, в которых косвенно содержится информация о целевой переменной (odometer/price_std, price).

Для обученной модели нужно:

рассчитать и вывести метрику качества (accuracy) на тренировочной выборке;
сделать предикт на тестовых данных и положить его в переменную logreg_pred;
рассчитать и вывести accuracy и confusion_matrix на тестовой выборке.
Задание засчитывается, если:

accuracy на тренировочной выборке > 87%;
accuracy на тестовой выборке > 75.5%.
Для улучшения метрики можно преобразовывать входные данные, тюнить параметры логистической регрессии.

In [12]:
df_log = df.copy()

In [13]:
df_log = df_log.drop(columns = ['odometer/price_std','price'])
df_log.head()

id price_category  is_audi  is_ford  is_chevrolet  is_toyota  \
0  7308295377           high        0        0             0          0   
1  7316380095         medium        0        1             0          0   
2  7313733749           high        0        0             0          0   
3  7308210929         medium        0        0             0          1   
4  7303797340         medium        0        0             0          0   

   x0_diesel  x0_electric  x0_gas  x0_hybrid  ...   lat_std  long_std  \
0        1.0          0.0     0.0        0.0  ... -0.619172  0.484245   
1        0.0          0.0     0.0        0.0  ...  0.388014  1.110800   
2        0.0          0.0     1.0        0.0  ... -0.843059  0.531185   
3        0.0          0.0     1.0        0.0  ... -0.504509  0.853562   
4        0.0          0.0     0.0        0.0  ... -0.460101  0.557607   

   year_std  desc_len_std  model_in_desc_std  model_len_std  \
0  1.322394      0.632075          -0.155788       1.163032   
1  0.695973     -0.646781          -0.155788       0.932087   
2  0.852578      0.560744          -0.155788       0.470197   
3  0.226157      0.180435          -0.155788      -0.915473   
4  0.069552      0.766366          -0.155788       1.163032   

   model_word_count_std  month_std  dayofweek_std  diff_years_std  
0              1.910669  -0.615846       1.120284       -1.322394  
1              1.235799   1.623784      -1.374972       -0.695973  
2              0.560930  -0.615846      -0.376870       -0.852578  
3             -0.788810  -0.615846       1.120284       -0.226157  
4              1.910669  -0.615846       0.122182       -0.069552  

[5 rows x 1461 columns]

In [14]:
x = df_log.drop(['price_category'], axis=1)
y = df.price_category

In [15]:
train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.3, random_state=42)

In [16]:
logreg = LogisticRegression()

In [17]:
logreg.fit(train_x, train_y)

LogisticRegression()

In [18]:
logreg_pred = logreg.predict(test_x)
accuracy_score(test_y, logreg_pred)

0.32432432432432434

In [19]:
confusion_matrix(test_y, logreg_pred)

array([[  0,   0, 997],
       [  0,   0, 953],
       [  0,   0, 936]], dtype=int64)

In [20]:
logreg_pred_2 = logreg.predict(train_x)
accuracy_score(train_y, logreg_pred_2)

0.3392247140947572

In [21]:
confusion_matrix(train_y, logreg_pred_2)

array([[   0,    0, 2202],
       [   0,    0, 2247],
       [   0,    0, 2284]], dtype=int64)

In [22]:
# приступаем к тюнингу 


In [23]:
df_log_new = df.copy()

In [24]:
df_log_new = df_log_new.drop((['id','odometer/price_std','price','manufacturer_dodge']), axis =1)
df_log_new.head()

price_category  is_audi  is_ford  is_chevrolet  is_toyota  x0_diesel  \
0           high        0        0             0          0        1.0   
1         medium        0        1             0          0        0.0   
2           high        0        0             0          0        0.0   
3         medium        0        0             0          1        0.0   
4         medium        0        0             0          0        0.0   

   x0_electric  x0_gas  x0_hybrid  x0_other  ...   lat_std  long_std  \
0          0.0     0.0        0.0       0.0  ... -0.619172  0.484245   
1          0.0     0.0        0.0       1.0  ...  0.388014  1.110800   
2          0.0     1.0        0.0       0.0  ... -0.843059  0.531185   
3          0.0     1.0        0.0       0.0  ... -0.504509  0.853562   
4          0.0     0.0        0.0       1.0  ... -0.460101  0.557607   

   year_std  desc_len_std  model_in_desc_std  model_len_std  \
0  1.322394      0.632075          -0.155788       1.163032   
1  0.695973     -0.646781          -0.155788       0.932087   
2  0.852578      0.560744          -0.155788       0.470197   
3  0.226157      0.180435          -0.155788      -0.915473   
4  0.069552      0.766366          -0.155788       1.163032   

   model_word_count_std  month_std  dayofweek_std  diff_years_std  
0              1.910669  -0.615846       1.120284       -1.322394  
1              1.235799   1.623784      -1.374972       -0.695973  
2              0.560930  -0.615846      -0.376870       -0.852578  
3             -0.788810  -0.615846       1.120284       -0.226157  
4              1.910669  -0.615846       0.122182       -0.069552  

[5 rows x 1459 columns]

In [25]:
x = df_log_new.drop(['price_category'], axis=1)
y = df_log_new.price_category

In [26]:
train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.3, random_state=60)

In [27]:
logreg = LogisticRegression(C = 1, penalty='none', max_iter = 200)
logreg.fit(train_x, train_y)

C:\Users\kish7\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(C=1, max_iter=200, penalty='none')

In [28]:
pred_train = logreg.predict(train_x)
pred_train

array(['low', 'low', 'low', ..., 'low', 'medium', 'medium'], dtype=object)

In [29]:
confusion_matrix(train_y, pred_train)

array([[2057,   15,  143],
       [  32, 2055,  183],
       [ 149,  205, 1894]], dtype=int64)

In [30]:
pred_test = logreg.predict(test_x)
pred_test

array(['medium', 'low', 'low', ..., 'low', 'medium', 'low'], dtype=object)

In [31]:
confusion_matrix(test_y, pred_test)

array([[806,  22, 156],
       [ 28, 742, 160],
       [138, 175, 659]], dtype=int64)

In [32]:
accuracy_score(train_y, pred_train)

0.8920243576414674

In [33]:
accuracy_score(test_y, pred_test)

0.7647262647262647

In [34]:
logreg.coef_

array([[ 0.69691732,  0.79007649,  1.19062539, ..., -0.02212196,
        -0.01877054, -1.17568264],
       [-0.42001077, -0.77768061, -0.9319664 , ...,  0.05224141,
         0.01205212,  1.02526222],
       [-0.27690655, -0.01239587, -0.25865899, ..., -0.03011944,
         0.00671842,  0.15042042]])

In [35]:
logreg.get_params()

{'C': 1,
 'class_weight': None,
 'dual': False,
 'fit_intercept': True,
 'intercept_scaling': 1,
 'l1_ratio': None,
 'max_iter': 200,
 'multi_class': 'auto',
 'n_jobs': None,
 'penalty': 'none',
 'random_state': None,
 'solver': 'lbfgs',
 'tol': 0.0001,
 'verbose': 0,
 'warm_start': False}

# Задача 2. Многослойный персептрон

Решите задачу, поставленную в предыдущем задании, применив модель многослойного персептрона. Добейтесь лучших значений точности на тренировочной и тестовой выборках, чем на логистической регрессии, с помощью данного алгоритма.

Значимым увеличением точности в данном задании будем считать увеличение больше, чем на 0.01 по метрике accuracy.

In [36]:
mlp = MLPClassifier(random_state=42, max_iter=500)
mlp.fit(train_x, train_y)

MLPClassifier(max_iter=500, random_state=42)

In [37]:
pred_mlp = mlp.predict(train_x)
accuracy_score(train_y, pred_mlp)

0.9998514777959305

In [38]:
pred_mlp_2 = mlp.predict(test_x)
accuracy_score(test_y, pred_mlp_2)

0.7747747747747747

In [39]:
mlp.get_params()

{'activation': 'relu',
 'alpha': 0.0001,
 'batch_size': 'auto',
 'beta_1': 0.9,
 'beta_2': 0.999,
 'early_stopping': False,
 'epsilon': 1e-08,
 'hidden_layer_sizes': (100,),
 'learning_rate': 'constant',
 'learning_rate_init': 0.001,
 'max_fun': 15000,
 'max_iter': 500,
 'momentum': 0.9,
 'n_iter_no_change': 10,
 'nesterovs_momentum': True,
 'power_t': 0.5,
 'random_state': 42,
 'shuffle': True,
 'solver': 'adam',
 'tol': 0.0001,
 'validation_fraction': 0.1,
 'verbose': False,
 'warm_start': False}

In [40]:
mlp.n_layers_

3

In [41]:
mlp_tuned = MLPClassifier(random_state=42, max_iter=500, hidden_layer_sizes=(100,20), activation = 'tanh')
mlp_tuned.fit(train_x, train_y)

MLPClassifier(activation='tanh', hidden_layer_sizes=(100, 20), max_iter=500,
              random_state=42)

In [42]:
mlp_tuned.n_layers_

4

In [43]:
pred_mlp3 = mlp_tuned.predict(train_x)
accuracy_score(train_y, pred_mlp3)

0.9992573889796524

In [44]:
pred_mlp4 = mlp_tuned.predict(test_x)
accuracy_score(test_y, pred_mlp4)

0.7813582813582813

Четырехслойная нейронная сеть показала лучший результат,чем логистическая регрессия 
тренировочная выборка: 
0.9992573889796524 > 0.8920243576414674
тестовая выборка:
0.7813582813582813 > 0.7647262647262647


# Задача 3. Сравнение с древовидными моделями

Обучите модель случайного леса на тех же данных для предсказания price_category. Сравните качество с моделью логистической регрессии и многослойного персептрона. Словами опишите, какая из моделей в каких случаях работает лучше по результатам на тестовой выборке, и обоснуйте свой выбор.

In [45]:
from sklearn.ensemble import RandomForestClassifier

In [46]:
x = df_log_new.drop(['price_category'], axis=1)
y = df_log_new.price_category
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=50)

In [47]:
rf_clf = RandomForestClassifier()
rf_clf.fit(x_train, y_train)

RandomForestClassifier()

In [48]:
predicted_train_rf = rf_clf.predict(x_train)
predicted_test_rf = rf_clf.predict(x_test)

In [49]:
print(accuracy_score(y_train, predicted_train_rf))
confusion_matrix(y_train, predicted_train_rf)

1.0


array([[2253,    0,    0],
       [   0, 2225,    0],
       [   0,    0, 2255]], dtype=int64)

In [50]:
print(accuracy_score(y_test, predicted_test_rf))
confusion_matrix(y_test, predicted_test_rf)

0.7633402633402634


array([[788,  44, 114],
       [ 13, 824, 138],
       [159, 215, 591]], dtype=int64)

На тестовой выборке лучше всех работает многослойный персептрон. Я думаю, что на результат повлияли max_iter=500, hidden_layer_sizes=(100,20) которые были установлены в нейронную сеть.

# Задача 4. Стратегия голосования

Реализуйте стратегию голосования для предсказания целевой переменной. Голосование в задаче классификации — это когда несколько моделей выдают свои предикты, и финальным выбирается тот предикт, который предсказали большинство моделей.

Для реализации этой стратегии проделайте следующее:

сохраните предсказания каждой из моделей (случайный лес, многослойный персептрон, логистическая регрессия) для тестовой выборки в датафрейм pred_df;

в четвертую колонку target положите тот класс, который предсказало большинство классификаторов. Например, если в строке были значения high, medium, medium, в target нужно положить medium.

Если в строке три разных класса — high, medium, low, — придумайте свою стратегию по выбору значения. Самая простая стратегия: выбрать рандомно одно значение из трёх.

посчитайте точность предсказания с помощью голосования. Выведите значения метрик accuracy и confusion_matrix.

Добейтесь значения точности > 78%

In [80]:
pred_df = pd.DataFrame({'Логистическая регрессия' : pred_test, 'Многослойный персептрон' : pred_mlp4, 'Случайный лес' : predicted_test_rf})
from collections import Counter
import random
def t(a,b,c):
    res = ' '
    tmp = {k:v for k,v in sorted (Counter([a,b,c]).items(),key = lambda item: item[1],reverse = True)}
    key = list(tmp.items())[0][0]
    value = list(tmp.items())[0][1]
    if not (value == 1):
        res = key
    else:
        res = random.choice([a,b,c])
    return res
pred_df['Target'] = pred_df.apply(lambda x: t(x['Случайный лес'],x['Многослойный персептрон'],x['Логистическая регрессия']),axis = 1)
pred_df

Логистическая регрессия Многослойный персептрон Случайный лес  Target
0                     medium                  medium          high  medium
1                        low                  medium          high    high
2                        low                  medium           low     low
3                        low                     low          high     low
4                       high                    high          high    high
...                      ...                     ...           ...     ...
2881                    high                    high           low    high
2882                     low                  medium          high  medium
2883                     low                     low        medium     low
2884                  medium                  medium        medium  medium
2885                     low                  medium          high    high

[2886 rows x 4 columns]

In [81]:
accuracy_score(y_test, pred_df['Target'])

0.37422037422037424

In [82]:
confusion_matrix(y_test, pred_df['Target'])

array([[370, 295, 281],
       [300, 370, 305],
       [336, 289, 340]], dtype=int64)